plugins/geofence/geofence_impl.cpp

```diff
@@ -0,0 +1,219 @@
+
+    // Reset the timeout because we're still communicating.
+    _parent->refresh_timeout_handler(this);
+}
+
+void GeofenceImpl::process_mission_ack(const mavlink_message_t &message)
+{
+    if (!_active) {
+        LogDebug() << "Ignore mission ack because not active";
+        return;
+    }
+
+    mavlink_mission_ack_t mission_ack;
+    mavlink_msg_mission_ack_decode(&message, &mission_ack);
+
+    if (mission_ack.target_system != _parent->get_own_system_id() &&
+        mission_ack.target_component != _parent->get_own_component_id()) {
+        LogDebug() << "Ignore geofence ack that is not for us";
+        return;
+    }
+
+    // We got some response, so it wasn't a timeout and we can remove it.
+    _parent->unregister_timeout_handler(this);
+
+    if (mission_ack.type == MAV_MISSION_ACCEPTED) {
+        report_geofence_result(_result_callback, Geofence::Result::SUCCESS);
+        LogDebug() << "Sucess, done";
+
+    } else if (mission_ack.type == MAV_MISSION_NO_SPACE) {
+        LogDebug() << "Error: too many geofence items: " << int(mission_ack.type);
+        report_geofence_result(_result_callback, Geofence::Result::TOO_MANY_GEOFENCE_ITEMS);
+    } else {
+        LogDebug() << "Error: unknown geofence ack: " << int(mission_ack.type);
+        report_geofence_result(_result_callback, Geofence::Result::ERROR);
+    }
+}
+
+void GeofenceImpl::assemble_mavlink_messages(
+    const std::vector<std::shared_ptr<Geofence::Polygon>> &polygons)
+{
+    // TODO: delete all entries first
+    _mavlink_geofence_item_messages.clear();
+
+    for (auto &polygon : polygons) {
+        uint16_t command;
+        switch (polygon->type) {
+            case Geofence::Polygon::Type::INCLUSION:
+                command = MAV_CMD_NAV_FENCE_POLYGON_VERTEX_INCLUSION;
+                break;
+            case Geofence::Polygon::Type::EXCLUSION:
+                command = MAV_CMD_NAV_FENCE_POLYGON_VERTEX_EXCLUSION;
+                break;
+            default:
+                LogDebug() << "Unknown type";
+                return;
+        }
+
+        for (auto &point : polygon->points) {
+            std::shared_ptr<mavlink_message_t> message(new mavlink_message_t());
+            mavlink_msg_mission_item_int_pack(_parent->get_own_system_id(),
+                                              _parent->get_own_component_id(),
+                                              message.get(),
+                                              _parent->get_system_id(),
+                                              _parent->get_autopilot_id(),
+                                              _mavlink_geofence_item_messages.size(),
+                                              MAV_FRAME_GLOBAL_INT,
+                                              command,
+                                              0, // current
+                                              0, // autocontinue
+                                              polygon->points.size(), // vertex count
```

Comment:
Some errors here and following lines:

> "C:\dronecode_sdk\build\plugins\geofence\dronecode_sdk_geofence.vcxproj" (default target) (11) ->
  C:\dronecode_sdk\plugins\geofence\geofence_impl.cpp(181): warning C4244: 'argument': conversion from 'double' to 'int32_t', possible loss of data [C:\dronecode_sdk\build\plugins\geofence\dronecode_sdk_geofence.vcxproj]
  C:\dronecode_sdk\plugins\geofence\geofence_impl.cpp(180): warning C4244: 'argument': conversion from 'double' to 'int32_t', possible loss of data [C:\dronecode_sdk\build\plugins\geofence\dronecode_sdk_geofence.vcxproj]
  C:\dronecode_sdk\plugins\geofence\geofence_impl.cpp(176): warning C4244: 'argument': conversion from 'unsigned __int64' to 'float', possible loss of data [C:\dronecode_sdk\build\plugins\geofence\dronecode_sdk_geofence.vcxproj]